### ISSUE TITLE
Cannot bind completion-at-point/corfu-complete and corfu-next to the same key

### ISSUE THREAD

**Issue by Overdr0ne** [minad/corfu#12]
First of all, thank you for this package, it seems to work pretty well ootb for the most part!

After binding a key to completion-at-point, e.g. TAB, when I then also try to bind that key to corfu-next in corfu-map, it doesn't work. Corfu seems to automatically bind that key to corfu-complete while the overlay is active. I do not see any use for an explicit corfu-complete function. When I cycle to the match I want with corfu-next/corfu-previous, I just continue typing and the overlay closes automatically, so I would rather not bind corfu-complete to an explicit key. Let me know if that is not clear.

thanks!

**Comment by minad** [comment]
You can bind `TAB` to `corfu-next` and `S-TAB` to `corfu-previous` in the `corfu-map`, which is activated when the popup is open. This is documented in the README, see https://github.com/minad/corfu#configuration. Maybe your problem is that the `corfu-map` also remaps `completion-at-point` to `corfu-complete`. You may want to disable that binding via `(define-key corfu-map [remap completion-at-point] nil)`. Note that `corfu-map` is just a normal keymap you can manipulate to your liking, Corfu does not manipulate that keymap. Does that answer your question?

**Comment by Overdr0ne** [comment]
Wow, thank you for getting back so fast.

After investigating further, the problem is not what I thought it was. I am using Evil, and when I switch to emacs mode, the binding for completion-at-point can also be used for corfu-next without issue. However, when I try to modify the corfu-map for the insert state, emacs does not see the new binding, but I can see the binding in corfu-map. In fact, evil seems to be overriding all bindings in corfu-map, except the [remap *] bindings. So it seems somehow the evil keymap is taking precedence over the corfu-map bindings.

So this is likely not your issue, but I will try to figure it out soon if you'd like to keep it open so I can comment. It's just strange because I've customized plenty of other keybindings for various minor modes for multiple evil states without a problem, so I don't know why it wouldn't work here...

**Comment by Overdr0ne** [comment]
What does this line do in corfu--setup? 
```
  (setcdr (assq #'completion-in-region-mode minor-mode-overriding-map-alist) corfu-map)
```
Do you happen to know if corfu-map can be made to take higher precedence when active?

**Comment by minad** [comment]
This line set the `minor-mode-overriding-map` of `completion-in-region-mode` to the `corfu-map`. The `completion-in-regio-mode` is the transient minor mode used for completion in region (i.e., when the corfu popup is open).

I close this issue then since this is an evil issue. I suggest you ask in the evil-collection repository for an advice on how to fix this.

Here you find the documentation about the keymap precedence: https://www.gnu.org/software/emacs/manual/html_node/elisp/Searching-Keymaps.html

~~~ elisp
(or (if overriding-terminal-local-map                      ;; hydra, transient, ...
        (find-in overriding-terminal-local-map))
    (if overriding-local-map
        (find-in overriding-local-map)
      (or (find-in (get-char-property (point) 'keymap))
          (find-in-any emulation-mode-map-alists)          ;; <- evil
          (find-in-any minor-mode-overriding-map-alist)    ;; <- corfu
          (find-in-any minor-mode-map-alist)               ;; minor modes
          (if (get-text-property (point) 'local-map)
              (find-in (get-char-property (point) 'local-map))
            (find-in (current-local-map)))))
    (find-in (current-global-map)))                              
~~~

**Comment by minad** [comment]
As an alternative I could also consider to bind corfu to the overriding maps, this is always a bit difficult to know where the correct location is. However I think the `minor-mode-overriding-map-alist` is exactly right since evil should actually take precedence over corfu, but maybe evil must be adjusted accordingly. For company you also need some small integration configuration (https://github.com/emacs-evil/evil-collection/blob/master/modes/company/evil-collection-company.el).

**Comment by Overdr0ne** [comment]
I don't understand the correct precendence either, but after appending emultion-mode-map-alists like below, it worked as expected. This seems to be what company does as well, so it might be appropriate. Or maybe it should be part of an evil-collection, not really sure...

```
@@ -552,7 +552,8 @@ If `line-spacing/=nil' or in text-mode, the background color is used instead.")
   ;; Keep completion alive when popup is shown (disable predicate check!)
   (remove-hook 'post-command-hook #'completion-in-region--postch)
   (setq corfu--extra-properties completion-extra-properties)
-  (setcdr (assq #'completion-in-region-mode minor-mode-overriding-map-alist) corfu-map)
+  (setq corfu-mode-map-alist `((completion-in-region-mode . ,corfu-map)))
+  (add-to-list 'emulation-mode-map-alists corfu-mode-map-alist)
   (add-hook 'pre-command-hook #'corfu--pre-command-hook nil 'local)
   (add-hook 'post-command-hook #'corfu--post-command-hook nil 'local))
```

**Comment by minad** [comment]
Yes, adding the map to the emulation map list is a possibility. But I don't think this is the correct fix since the Emacs built-in `completion-in-region-mode` puts its maps into `minor-mode-overriding-map-alist`. The goal here is to stay close to the built-ins. Maybe the workaround can go to evil-collection or you can simply add it to your init.el. Adding it to evil-collection is a bit early since Corfu is very new and things may still change.

**Comment by Stebalien** [comment]
Unfortunately, the "correct" way to do this is to either:

1. Call `evil-normalize-keymaps` after modifying `minor-mode-overriding-alist`.
2. Define the keys with `evil-define-minor-mode-key`.

The first one requires either advising or modifying corfu and the second one is annoying, to say the least, because the keys need to be attached to a minor _mode_, not a map. Given that `corfu-mode` is always active, the keys need to be attached to `completion-in-region-mode` (which applies globally...). Yes, attached to the _mode_, not attached to a map (evil handles the map at that point).

```elisp
(evil-define-minor-mode-key 'insert 'completion-in-region-mode
 ...
```

**Comment by Stebalien** [comment]
E.g., with general:

```elisp
(general-def
 :keymaps 'completion-in-region-mode
 :definer 'minor-mode
 :states 'insert
 :predicate 'corfu-mode
 "<tab>" 'corfu-next
 "<backtab>" 'corfu-previous
 "C-j" 'corfu-next
 "C-k" 'corfu-previous
 "C-l" 'corfu-complete
 "<escape>" 'corfu-abort
 "<return>" 'corfu-insert)
 ```

**Comment by minad** [comment]
Why not advise corfu? Seems like the better solution. This could be added to evil-collection and I can add a note to the readme.

**Comment by Stebalien** [comment]
`evil-normalize-keymaps` is a non-trivial function and I wanted to try to figure out the "right" way to do this.

But yeah, the advising method is probably more general.

Specifically, you need:

```elisp
  (evil-make-overriding-map corfu-map)
  (advice-add 'corfu--setup :after 'evil-normalize-keymaps)
  (advice-add 'corfu--teardown :after 'evil-normalize-keymaps)
```

**Comment by minad** [comment]
@Stebalien And what about doing the keymap adjustment described by @Overdr0ne in the advice of `corfu--setup`? This would be simpler than calling `evil-normalize-keymaps`.

**Comment by Stebalien** [comment]
That's definitely a simpler solution, although I'm not sure which is considered to be "more correct".

**Comment by Overdr0ne** [comment]
Hey, I should have responded before. I don't think what I posted before actually works. I realized later that in my testing, I had just somehow globally overwritten those keybindings, instead of just for the insert state. I'm still not sure what the really right way to fix this is, or why adding the corfu-map to the emulation-mode-map-alist didn't work. Even as much as I use it, I just don't grok Evil's architecture very well. What Stebalien says makes sense to me, and I need to study it more, but what I ended up doing was just disabling Evil's default insert state keybindings with (setq evil-disable-insert-state-bindings t), then rebound what I wanted from them with General, which allows me to fine tune the insert state for each mode. I believe General prepends <insert-state> to the keybinding instead of adding it to the evil-insert-state-map, and those flags are handled specially somewhere in the command loop. Pretty ugly, but it bypasses the emulation-mode-map-alist precedence issue.

tl;dr I think Stebalien's solution is probably the best bet for the average Evil user, my setup is a bit unusual.

**Comment by minikN** [comment]
Hello,

Is there a way to make this work when invoking `dabbrev-completion`? 

```lisp
  (evil-make-overriding-map corfu-map)
  (advice-add 'corfu--setup :after 'evil-normalize-keymaps)
  (advice-add 'corfu--teardown :after 'evil-normalize-keymaps)
```
This works. However if I display corfu by invoking `dabbrev-completion` it still bugs. 
`C-j` and `C-k` invoke org-mode commands instead of `corfu-next` and `corfu-previous`.
This only fails when in insert mode, it works in normal mode.

My config:

```lisp
(use-package! corfu
  :bind (:map corfu-map
         ("C-j" . corfu-next)
         ("C-k" . corfu-previous))
  :custom
  (corfu-cycle t)
  (corfu-auto t)

  :init
  (corfu-global-mode)

  :config
  ;; https://github.com/minad/corfu/issues/12#issuecomment-869037519
  (advice-add 'corfu--setup :after 'evil-normalize-keymaps)
  (advice-add 'corfu--teardown :after 'evil-normalize-keymaps)
  (evil-make-overriding-map corfu-map))
```

**Comment by Overdr0ne** [comment]
> This works. However if I display corfu by invoking dabbrev-completion it still bugs.
C-j and C-k invoke org-mode commands instead of corfu-next and corfu-previous.
This only fails when in insert mode, it works in normal mode.

Hmm, so this happens in org mode, in the insert state? Does calling evil-normalize-keymaps manually when corfu is active make it work? How about (evil-normalize-keymaps 'insert)?

**Comment by minikN** [comment]
> > This works. However if I display corfu by invoking dabbrev-completion it still bugs.
> > C-j and C-k invoke org-mode commands instead of corfu-next and corfu-previous.
> > This only fails when in insert mode, it works in normal mode.
> 
> Hmm, so this happens in org mode, in the insert state? Does calling evil-normalize-keymaps manually when corfu is active make it work? How about (evil-normalize-keymaps 'insert)?

Hm maybe, if you tell me how to execute the function while corfu is active :)
With __active__ I mean the corfu popup is visible. I can't to `M-x` or `M-:` while it is active, it will make it disappear. If I invoke corfu again after running `(evil-normalize-keymaps 'insert)` it doesn't make a difference.

**Comment by zrquan** [comment]
@minikN are you using doomemacs? You can use `evil-org-set-key-theme` to disable the insert keytheme of evil-org-mode, this work for me

**Comment by drcxd** [comment]
I got some problem with this piece of code:
```
  (evil-make-overriding-map corfu-map)
  (advice-add 'corfu--setup :after 'evil-normalize-keymaps)
  (advice-add 'corfu--teardown :after 'evil-normalize-keymaps)
```
My corfu configuration is:
```
(setq corfu-auto t
      corfu-cycle nil
      corfu-quit-at-boundary nil
      corfu-quit-no-match t
      corfu-preview-current nil)

(define-key corfu-map "\C-p" #'corfu-previous)
(define-key corfu-map "\C-n" #'corfu-next)
(define-key corfu-map (kbd "<escape>") #'corfu-quit)
```
My intention is to override the default evil map so I can use `C-n` and `C-p` to traverse completion candidates. This works fine at the first glance. However, I found that with my configuration, after I insert some candidates by typing enter, i.e. `corfu-insert`, the `corfu-map` overriding is still taking effect, though there is no candidates prompting. As a result, I found myself mysteriously typing enter but cannot input the newline character, simply because enter is still bound to `corfu-insert`. Any idea how to fix this problem?

**Comment by Vurp** [comment]
The suggestions described in this thread doesn't work. I think this issue should be reopened.
